File: mesh.nim
import nimgl/opengl

type
  Mesh* = object
    vao*: uint32
    vbo*: uint32
    idx*: uint32
    len: int32

proc newMesh*(vertices: var seq[float32], indices: var seq[uint32]): Mesh =
  glGenVertexArrays(1, result.vao.addr)
  glGenBuffers(1, result.vbo.addr)
  glGenBuffers(1, result.idx.addr)
  glBindVertexArray(result.vao)

  if indices.len == 0:
    for i in 0 ..< (vertices.len / 3).int32:
      indices.add(i.uint32)

  glBindBuffer(GL_ARRAY_BUFFER, result.vbo)
  glBufferData(GL_ARRAY_BUFFER, float32.sizeof * vertices.len, vertices[0].addr, GL_STATIC_DRAW)

  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, result.idx)
  glBufferData(GL_ELEMENT_ARRAY_BUFFER, int32.sizeof * indices.len, indices[0].addr, GL_STATIC_DRAW)
  result.len = indices.len.int32

  glVertexAttribPointer(0, 3, EGL_FLOAT, false, 5 * float32.sizeof(), nil)
  glEnableVertexAttribArray(0)
  glVertexAttribPointer(1, 2, EGL_FLOAT, false, 5 * float32.sizeof(), cast[pointer](float32.sizeof * 3))
  glEnableVertexAttribArray(1)

proc destroy*(mesh: var Mesh) =
  glDeleteVertexArrays(1, mesh.vao.addr)
  glDeleteBuffers(1, mesh.vbo.addr)
  glDeleteBuffers(1, mesh.idx.addr)

proc draw*(mesh: Mesh) =
  glBindVertexArray(mesh.vao)
  glDrawElements(GL_TRIANGLES, mesh.len, GL_UNSIGNED_INT, nil)

File: mesh.cc
#include "mesh.h"

#include "opengl.h"

namespace vx {

Mesh::Mesh(const std::vector<f32> &vertices, const std::vector<u32> &indices):
vertices_count_(vertices.size()), index_count_(indices.size()) {
  glGenBuffers(2, vbo_);

#ifndef __EMSCRIPTEN__
  glGenVertexArrays(1, &vao_);
  glBindVertexArray(vao_);
#endif

  glBindBuffer(GL_ARRAY_BUFFER, vbo_[0]);
  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, vbo_[1]);

  glBufferData(GL_ARRAY_BUFFER, sizeof(f32) * vertices_count_, &vertices[0], GL_STATIC_DRAW);
  glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(u32) * index_count_, &indices[0], GL_STATIC_DRAW);

  glEnableVertexAttribArray(0);
  glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, sizeof(f32) * 5, (void *)0);
  glEnableVertexAttribArray(1);
  glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, sizeof(f32) * 5, (void *)(3 * sizeof(f32)));

  glBindVertexArray(0);
  glBindBuffer(GL_ARRAY_BUFFER, 0);
  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
}

Mesh::~Mesh() {
  glDeleteVertexArrays(1, &vao_);
  glDeleteBuffers(2, vbo_);
}

void Mesh::draw() {
  glBindVertexArray(vao_);
  glBindBuffer(GL_ARRAY_BUFFER, vbo_[0]);
  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, vbo_[1]);

  glDrawElements(GL_TRIANGLES, index_count_, GL_UNSIGNED_INT, (void *)0);

  glBindVertexArray(0);
}

}
